### PR TITLE
Bug Fix: pd.DataFrame.sum with min_count changes dtype if result contains NaNs

### LINKED ISSUE #46947: BUG: pd.DataFrame.sum with min_count changes dtype if result contains NaNs
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
>>> import numpy as np
>>> import pandas as pd
>>> df = pd.DataFrame({'a': [1., 2.3, 4.4], 'b': [2.2, 3, np.nan]}, dtype='float32')
>>> df.sum(axis=1, min_count=1).dtype
dtype('float32')  # OK
>>> df.sum(axis=1, min_count=2).dtype
dtype('float64')  # not OK
>>> df.sum(axis=1, skipna=False).dtype
dtype('float32')  # OK
```

### Issue Description

When using min_count, the dtype of the result depends on whether the result contains a NaN or not.

### Expected Behavior

The resulting dtype should be the same as the dtype of the original DataFrame, also when the result contains NaNs.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 4bfe3d07b4858144c219b9346329027024102ab6
python           : 3.10.2.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19043
machine          : AMD64
processor        : Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United Kingdom.1252
pandas           : 1.4.2
numpy            : 1.22.3
pytz             : 2022.1
dateutil         : 2.8.2
pip              : 22.0.4
setuptools       : 58.1.0
Cython           : None
pytest           : 7.0.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 3.0.3
lxml.etree       : 4.8.0
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.1
IPython          : 8.1.0
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : 1.3.4
brotli           : None
fastparquet      : None
fsspec           : 2022.02.0
gcsfs            : None
markupsafe       : 2.1.0
matplotlib       : 3.5.1
numba            : None
numexpr          : 2.8.1
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 7.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.8.0
snappy           : None
sqlalchemy       : 1.4.32
tables           : None
tabulate         : 0.8.9
xarray           : 2022.3.0
xlrd             : None
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by weikhor**
- [ ] closes #46947
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/frame/test_reductions.py [hunk 1]
```diff
@@ -778,6 +778,20 @@ def test_sum_nanops_min_count(self):
...
+            {"axis": 1, "min_count": 1},
+            {"axis": 1, "min_count": 2},
+            {"axis": 1, "skipna": False},
+        ],
+    )
+    def test_sum_nanops_dtype_min_count(self, kwargs):
+        # GH#46947
+        df = DataFrame({"a": [1.0, 2.3, 4.4], "b": [2.2, 3, np.nan]}, dtype="float32")
+        result = df.sum(**kwargs).dtype
+        assert result == "float32"
```
Comment (phofl) [inline pandas/tests/frame/test_reductions.py:793]: Please test the whole DataFrame and parametrize over all possible float dtypes.

File: pandas/core/nanops.py [hunk 2]
```diff
@@ -1472,7 +1472,8 @@ def _maybe_null_out(
                 if np.iscomplexobj(result):
                     result = result.astype("c16")
                 else:
-                    result = result.astype("f8")
+                    if not is_float_dtype(result):
+                        result = result.astype("f8")
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/nanops.py:1476]: add `copy=False`

File: doc/source/whatsnew/v1.5.0.rst [hunk 3]
```diff
@@ -780,7 +780,7 @@ Indexing
 - Bug in setting large integer values into :class:`Series` with ``float32`` or ``float16`` dtype incorrectly altering these values instead of coercing to ``float64`` dtype (:issue:`45844`)
 - Bug in :meth:`Series.asof` and :meth:`DataFrame.asof` incorrectly casting bool-dtype results to ``float64`` dtype (:issue:`16063`)
 - Bug in :meth:`NDFrame.xs`, :meth:`DataFrame.iterrows`, :meth:`DataFrame.loc` and :meth:`DataFrame.iloc` not always propagating metadata (:issue:`28283`)
--
+- Bug in  :meth:`_maybe_null_out` pd.DataFrame.sum with min_count changes dtype if result contains NaNs
```
[the remark this answers is not in the document]
Comment (weikhor) [reply doc/source/whatsnew/v1.5.0.rst:783]: ok Have updated

**Comment by jreback** [comment]
thanks @weikhor very nice. can you add similar tests for `.prod()` (which i think should just work).

**Pull request merged**